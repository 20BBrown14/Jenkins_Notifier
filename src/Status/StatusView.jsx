import React from 'react';
import PropTypes from 'prop-types';
import Button from 'terra-button';
import Arrange from 'terra-arrange/lib/Arrange';
import './StatusView.scss';
import Text from 'terra-text';

const propTypes = {
  /** Repo information stored in app */
  repos: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  /** Button click handler for add new repo button */
  addRepoClickHandler: PropTypes.func.isRequired,
  /** Function to remove repo from list */
  removeRepo: PropTypes.func.isRequired,
};

const StatusView = (props) => {
  const {
    repos,
    addRepoClickHandler,
    removeRepo,
  } = props;

  const keys = [];
  if (repos) {
    Object.keys(repos).forEach((key) => {
      keys.push(key);
    });
  }

  /*
  for (let key in repos) {
    if (!repos.hasOwnProperty(key)) continue;
    keys.push(key);
  }
  */

  const content = [];
  for (let i = 0; i < keys.length; i += 1) {
    const arrangeFill = (
      <Text
        fontSize={16}
      >
        {keys[i]}
      </Text>);

    const arrangeStart = (
      <div>
        <Button
          key={`${keys[i]} start Jenkins`}
          className="arrangeButton"
          text="View on Jenkins"
          onClick={() => { window.open(repos[keys[i]].URL); }}
        />
        <Button
          key={`${keys[i]} start jobs`}
          className="arrangeButton"
          text="View Jobs"
          onClick={() => {}}
        />
      </div>
    );

    const arrangeEnd = (
      <Button
        key={`${keys[i]} end`}
        className="arrangeButton"
        text="Remove Repo"
        onClick={() => { removeRepo(`${keys[i]}`); }}
      />
    );
    content.push(
      <Arrange
        key={`${keys[i]} fill`}
        className="statusArrange"
        fitStart={arrangeStart}
        alignFitStart="center"
        fill={arrangeFill}
        alignFill="center"
        fitEnd={arrangeEnd}
        alignFitEnd="center"
      />,
    );
  }
  // content.push(<Button style={{ margin: '0px 5px 0px 5px' }} text="Confirm" />);

  return (
    <div>
      <div className="StatusView-Window">
        {content}
      </div>
      <Button
        style={{ margin: '12.5px 5px 0px 5px' }}
        text="Add New Repo"
        onClick={addRepoClickHandler}
      />
    </div>
  );
};

StatusView.propTypes = propTypes;
export default StatusView;
