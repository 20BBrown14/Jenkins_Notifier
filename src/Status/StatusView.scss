.StatusView-Window {
  width: 650px;
  height: 370px;
  overflow-y: auto;
  border-bottom: inset;
  border-bottom-color: lightgray;
  border-width: 2px;
}

.statusArrange {
  border-bottom: solid;
  border-width: 2px;
  border-color: lightgray;
}

.arrangeButton {
  margin: 0px;
}

.center {
  margin: auto;
  width: 100%;
  padding: 5px;
}

.titleText {
  margin: auto;
  width: 100%;
  padding: 5px;
  border-bottom: solid;
  border-width: 2px;
  border-color: lightgray;
  text-align: center;
}

.textCenter {
  text-align: center;
}

.tooltip {
  position: relative;
  display: inline-block;
}

.tooltip .tooltiptext {
  visibility: hidden;
  width: 120px;
  background-color: #808080;
  text-align: center;
  border-radius: 6px;
  padding: 5px 0;
  position: absolute;
  z-index: 1;
  top: 85%;
  opacity: 0;
  transition: opacity 0.5s;
  transition-delay: 0.75s;
}

.tooltip:hover .tooltiptext{
  visibility: visible;
  opacity: 1;
}

.refreshButton {
  top: -22px;
  right: 5px;
  position: absolute;
}

.addRepoButton {
  position: relative;
  float: right;
  top: 5px;
}