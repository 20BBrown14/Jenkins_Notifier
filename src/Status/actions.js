/** @const action for when Add New Repo button is clicked */
export const A_ADD_NEW_REPO_CLICKED = 'add_new_repo_clicked';
/** @const action to remove repo from list */
export const A_REMOVE_REPO = 'remove_repo';
/** @const action to view a repo's jobs */
export const A_VIEW_JOBS_CLICKED = 'view_jobs_clicked';
/** @const action to remove job from repo */
export const A_REMOVE_JOB_CLICKED = 'remove_job_clicked';
/** @const action to go back to repo view from jobs view */
export const A_GO_BACK_TO_REPO_VIEW = 'go_back_to_repo_view';


export const removeRepoAction = repoToRemove => (
  {
    type: A_REMOVE_REPO,
    data: { repoToRemove },
  }
);


export const viewJobsClickedAction = repoToView => (
  {
    type: A_VIEW_JOBS_CLICKED,
    data: { repoToView },
  }
);

export const removeJobAction = (jobToRemove, repo) => (
  {
    type: A_REMOVE_JOB_CLICKED,
    data: { jobToRemove, repo },
  }
);
