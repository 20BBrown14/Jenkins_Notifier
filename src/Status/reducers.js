import { A_VIEW_JOBS_CLICKED, A_GO_BACK_TO_REPO_VIEW } from './actions';

export const STATUS_STATE = 'status state';

const INITIAL_STATE = {
  repoToView: undefined,
};

const reduceViewJobsClicked = (state, action) => (
  {
    ...state,
    repoToView: action.data.repoToView,
  }
);

/**
 * Define Status reducers
 * @param {object} state - State with default value of INITIAL_STATE
 * @param {object} action - The action being performed
 * @param {string} action.type - The type of action being performed
 * @param {object} action.data - The data related to the action
 * @returns {object} updated state. If no action is handled returns original state.
 */
const reduceStatus = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case A_VIEW_JOBS_CLICKED:
      return reduceViewJobsClicked(state, action);
    case A_GO_BACK_TO_REPO_VIEW:
      return { ...state, repoToView: undefined };
    default:
      return state;
  }
};

export default { [STATUS_STATE]: reduceStatus };
