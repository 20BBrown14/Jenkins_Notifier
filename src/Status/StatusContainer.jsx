import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import StatusView from './StatusView';
import { A_ADD_NEW_REPO_CLICKED, removeRepoAction } from './actions';
import { APP_STATE } from '../reducers';

const propTypes = {
  /** Repo information stored */
  repos: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  /** Function for when add new repo button is clicked */
  addNewRepoClicked: PropTypes.func.isRequired,
  /** Function for when no repos are in the state when the app first loads */
  noRepos: PropTypes.func.isRequired,
  /** Function to remove repo from app state */
  removeRepo: PropTypes.func.isRequired,
};

export class StatusContainer extends React.Component {
  componentDidMount() {
    const {
      repos,
      noRepos,
    } = this.props;
    if (!repos) {
      noRepos();
    }
  }

  componentDidUpdate() {
    const {
      repos,
      noRepos,
    } = this.props;
    if (!repos) {
      noRepos();
    }
  }

  render() {
    const {
      repos,
      addNewRepoClicked,
      removeRepo,
    } = this.props;
    return (
      <div>
        <StatusView
          repos={repos}
          addRepoClickHandler={addNewRepoClicked}
          removeRepo={removeRepo}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  repos: state[APP_STATE].repos,
});

export const mapDispatchToProps = dispatch => ({
  addNewRepoClicked: () => {
    dispatch({ type: A_ADD_NEW_REPO_CLICKED });
  },
  noRepos: () => {
    dispatch({ type: A_ADD_NEW_REPO_CLICKED });
  },
  removeRepo: (repoToRemove) => {
    dispatch(removeRepoAction(repoToRemove));
  },
});

StatusContainer.propTypes = propTypes;
export default connect(mapStateToProps, mapDispatchToProps)(StatusContainer);
