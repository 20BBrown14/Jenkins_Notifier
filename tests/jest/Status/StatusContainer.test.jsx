import React from 'react';
import { StatusContainer, mapDispatchToProps } from '../../../src/Status/StatusContainer';
import {
  A_ADD_NEW_REPO_CLICKED,
  A_REMOVE_REPO, A_VIEW_JOBS_CLICKED,
  A_REMOVE_JOB_CLICKED,
  A_GO_BACK_TO_REPO_VIEW,
} from '../../../src/Status/actions';

describe('StatusContainer', () => {
  describe('Initialization', () => {
    it('should render a default view', () => {
      const testContainer = (shallow(
        <StatusContainer
          repos={{}}
          addNewRepoClicked={() => {}}
          noRepos={() => {}}
          removeRepo={() => {}}
          viewJobs={() => {}}
          removeJob={() => {}}
          goBackToRepoView={() => {}}
        />,
      ));
      expect(testContainer).toMatchSnapshot();
    });
  });
  describe('Action dispatch', () => {
    let mockDispatch;
    afterEach(() => {
      expect(mockDispatch).toHaveBeenCalled();
      mockDispatch = undefined;
    });
    it('dispatches an add new repo clicked action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_ADD_NEW_REPO_CLICKED);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.addNewRepoClicked();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches an add new repo clicked action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_ADD_NEW_REPO_CLICKED);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.noRepos();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches a remove repo action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_REMOVE_REPO);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.removeRepo();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches a view jobs action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_VIEW_JOBS_CLICKED);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.viewJobs('someRepo');
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches a remove jobs action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_REMOVE_JOB_CLICKED);
        expect(action.data.repo).toEqual('someRepoView');
        expect(action.data.jobToRemove).toEqual('someJobToRemove');
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.removeJob('someJobToRemove', 'someRepoView');
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches a go back action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_GO_BACK_TO_REPO_VIEW);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.goBackToRepoView();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
  });
  describe('Component Did Mount', () => {
    it('should call noRepos when there are no repos in state', () => {
      const wrapper = shallow(
        <StatusContainer
          repos={null}
          addNewRepoClicked={() => {}}
          noRepos={jest.fn()}
          removeRepo={() => {}}
          viewJobs={() => {}}
          removeJob={() => {}}
          goBackToRepoView={() => {}}
        />,
      );
      expect(wrapper.instance().props.noRepos).toHaveBeenCalled();
      expect(wrapper.instance().props.noRepos).toHaveBeenCalledTimes(1);
    });
  });

  describe('Component did update', () => {
    it('should call noRepos when there are no repos in state', () => {
      const wrapper = shallow(
        <StatusContainer
          repos={null}
          addNewRepoClicked={() => {}}
          noRepos={() => {}}
          removeRepo={() => {}}
          viewJobs={() => {}}
          removeJob={() => {}}
          goBackToRepoView={() => {}}
        />,
      );
      wrapper.setProps({ repos: undefined, noRepos: jest.fn() });
      expect(wrapper.instance().props.noRepos).toHaveBeenCalled();
      expect(wrapper.instance().props.noRepos).toHaveBeenCalledTimes(1);
    });
  });
});
