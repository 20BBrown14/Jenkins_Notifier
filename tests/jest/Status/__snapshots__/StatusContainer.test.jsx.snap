// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusContainer Initialization should render a default view 1`] = `
<div>
  <StatusView
    addRepoClickHandler={[Function]}
    goBack={[Function]}
    isLoading={false}
    refreshRepo={[Function]}
    removeJob={[Function]}
    removeRepo={[Function]}
    repos={Object {}}
    viewJobs={[Function]}
  />
</div>
`;
